Convert alerts directive to component and add spec test

TheRealJon opened the pull request:
## Description
Converted Angular directive `app/scripts/directives/alerts.js` into component at `app/components/alerts/alerts.component.js`. Added a spec test to the test suite which unit tests the new alerts component.

## Changes
- Create alerts component
- Add call to `close(alert)` function when an alert's link `onClick` function returns true. This ensures that the alert's `onClose` callback is called when it is dismissed by a link.
- Refactor alert template for use in component rather than directive.
- Remove alerts directive
- Load component in index.html
- Add spec test for component
- Add external html template to javascript module conversion to karma config (so that external templates can be loaded in spec tests without the http request).
- Load external template module in `spec-helper.js`
- Move alert component, template, and spec test files to `app/components/alerts` folder
- Adjust grunt and karma config for new alerts component file locations.
- Remove alerts component `toast` property
- Remove alerts component `animateSlide` property
- Change `filter` and `hideCloseButton` properties to one-way data binding in alerts component

benjaminapetersen commented on app/scripts/components/alerts.component.js:
```diff
@@ -0,0 +1,45 @@
+'use strict';
```
I'm a fan of moving things to a `scripts/components` directory, but would like to see a step further.  Can we do something like:

```
- scripts 
   - components 
     - alerts
        - alert.component.js
        - alert.view.html
```
I'm 150% into a component based heirarchy, its so much easier to derive relationships.  I realize this will require a little bit of tweaking to some of our build tools, but I doubt its much more than tweaking the globbing patterns.

@spadgett @jwforres may not agree, so I'll defer to their preference.

TheRealJon replied:
I agree, I would probably go even one step further and get rid of the scripts folder:
```
app/
  components/
    alerts/
      alerts.component.js
      alerts.component.spec.js
      alerts.template.html
```
Also, should we think about breaking this out into a separate story? I feel like introducing file structure changes in increments that are too small (one component at a time) will be confusing.

benjaminapetersen replied:
Agree on pulling it out of `scripts` since components are not 100% scripts.

We def prefer small PRs over big sweeping PRs, so I don't think this is a problem.  Rebases of giant structural PRs are terribly painful.

TheRealJon replied:
Yeah, I was thinking it could still be incremental, just more than one file at a time. We could choose whole sections to do at a time and break it up into a few PRs.

spadgett replied:
It seems like a good time to do this.

benjaminapetersen commented on app/scripts/components/alerts.component.js:
```diff
@@ -0,0 +1,45 @@
+angular.module('openshiftConsole')
+  .component('alerts', {
+    bindings: {
+      // Map (or array) of alert objects, each with properties
```
Maybe just put an exampe of the object here:

```JavaScript
alerts: {
  type: '',            // text....
  message: '',    // text....
  links: [{
     href: ''           // text....
     // etc.
  }],
  onClose: ''      // text....
}
```
Might be easier to consume

benjaminapetersen commented on app/scripts/components/alerts.component.js:
```diff
@@ -0,0 +1,45 @@
+      //   onClick:  handler to call when the link if clicked (optional)
+      //             If onClick returns true, dismisses the alert
+      alerts: '=',
+      // Filter function (optional).
```
`Filter function` is a bit too obvious of a comment here.  Can you update it to describe what/why this would be used?

TheRealJon replied:
I can clarify what this filter function does. Just realized that this probably needs a spec test case as well.

benjaminapetersen commented on app/scripts/components/alerts.component.js:
```diff
@@ -0,0 +1,45 @@
+
+      ctrl.onClick = function(alert, link) {
+        if (_.isFunction(link.onClick)) {
+          // If onClick() returns true, also hide the alert.
```
Which `onClick` returns `true`?  `ctrl.onClick` or `links.onClick`?  This comment is a bit confusing.

benjaminapetersen replied:
(preference for no comment if the code can be altered a bit to make that clear)

TheRealJon replied:
Most of the comments were just copy/paste from directive, but I can go back and spruce them up if needed.

spadgett replied:
Can you update the comment to say `link.onClick()`?

I think the comment here is helpful.

benjaminapetersen commented on app/scripts/components/alerts.component.js:
```diff
@@ -0,0 +1,45 @@
+      toast: '=?'
+    },
+    templateUrl: 'views/_alerts.html',
+    controller: function() {
```
We have a convention for using named functions for controllers for components.  That said, I prefer inline anonymous, I don't think we gain anything out of the named functions other than the need for IIFE's around all of these files.  

@spadgett I think may have a preference for the named function however, so I defer to his call on it.

benjaminapetersen commented on app/views/_alerts.html:
```diff
@@ -1,28 +1,16 @@
-      'mar-left-sm': toast
-    }">
-      <button ng-if="!hideCloseButton" ng-click="close(alert)" type="button" class="close">
+<div ng-attr-row="{{$ctrl.toast}}"  ng-attr-wrap="{{$ctrl.toast}}">
```
See above comment, but I'd love to see this file moved to `app/scripts/components/alerts/alert.view.html`.  Colocation is 😄  😀

benjaminapetersen commented on app/views/_alerts.html:
```diff
@@ -1,28 +1,16 @@
-      <button ng-if="!hideCloseButton" ng-click="close(alert)" type="button" class="close">
+<div ng-attr-row="{{$ctrl.toast}}"  ng-attr-wrap="{{$ctrl.toast}}">
+  <div ng-repeat="(alertID, alert) in ($ctrl.alerts | filterCollection : filter) track by (alertID + (alert.message || alert.details))" ng-if="!alert.hidden" class="alert-wrapper animate-repeat" ng-class="{'animate-slide': $ctrl.animateSlide}">
+    <div class="alert word-break" ng-class="{ 'alert-danger': alert.type === 'error', 'alert-warning': alert.type === 'warning', 'alert-success': alert.type === 'success', 'alert-info': !alert.type || alert.type === 'info', 'toast-pf': $ctrl.toast, 'mar-left-sm': $ctrl.toast}">
```
Not sure if we have a consensus on style in the console, but I def prefer the previous style:

```html 
<!-- preferred -->
<div
   id="foo"
   class="bar"
   ng-complicated-thing="break,it,up,again"></div>

<!-- not preferred, paragraph form gets hard to parse quickly -->
<div id="foo" class="bar" ng-complicated-thing="break,it,up,again"></div>
```
That and extremely long line lengths make it difficult to use multiple columns in a text editor.  Its nice to open JS `controller` next to HTML `view` and sometimes a JS `service` and be able to see everything.

We don't have a style guide tho, so I'll defer again to @jwforres @spadgett!

TheRealJon replied:
I do like the idea of breaking these long lines down, but for some reason, it breaks the spec test. The compiled element does not get the proper class applied when it's formatted that way.

benjaminapetersen replied:
oh it breaks the test?  what output do you get?

benjaminapetersen commented on test/spec/components/alerts.component.spec.js:
```diff
@@ -0,0 +1,99 @@
+        },
+        warning: {
+          type: "warning",
+          message: "I'm going to count to five."
```
hehe....

benjaminapetersen commented on test/spec/components/alerts.component.spec.js:
```diff
@@ -0,0 +1,99 @@
+  }));
+
+  // Ensure four alert elements exist
+  it("contain alert elements", function () {
```
We have had language like `it('should contain....')`.  Dunno if we care about that `should` or not.

spadgett replied:
+1, I prefer having the "should" in there

benjaminapetersen commented on test/spec/components/alerts.component.spec.js:
```diff
@@ -0,0 +1,99 @@
+    expect(allAlerts.length).toBe(4); // four alerts exist
+  });
+
+  // Ensure danger alert exists and is correct
```
Generally pretty good with tests.  Do we need to test negative cases along with positive cases?  Example, `warning` should not have `alert-success`, etc.  That could be a bit tedious for this, tho I think its good practice to test both sides.

benjaminapetersen commented on test/spec/components/alerts.component.spec.js:
```diff
@@ -0,0 +1,99 @@
+  // Ensure `close` function is called when close button is clicked, and the
+  // hidden state is true.
+  it("call contoller close function", function(){
+    $scope.onCloseSpy = jasmine.createSpy('onCloseSpy');
```
+1 to using a `spy`, I think this may be the first in our tests?

TheRealJon replied:
Just noticed a typo there too... "call contoller close function"

spadgett commented on app/scripts/components/alerts.component.js:
```diff
@@ -0,0 +1,45 @@
+      //   label:    the link text
+      //   onClick:  handler to call when the link if clicked (optional)
+      //             If onClick returns true, dismisses the alert
+      alerts: '=',
```
I think we should switch these to `<` bindings unless there is a reason not to.

(Not sure if this causes a problem with how we handle close since we set the hidden property inside the component. We might need to leave `alerts` as a `=` binding.)

spadgett commented on app/scripts/components/alerts.component.js:
```diff
@@ -0,0 +1,45 @@
+      filter: '=?',
+      animateSlide: '=?',
+      hideCloseButton: '=?',
+      toast: '=?'
```
I think we can remove this now. @TheRealJon can you sweep the code to check if it's still used anywhere?

Can you also check `animateSlide`?

spadgett commented on test/spec/components/alerts.component.spec.js:
```diff
@@ -0,0 +1,99 @@
+  beforeEach(inject(function(_$compile_, _$rootScope_) {
+      $scope = _$rootScope_;
+      $compile = _$compile_;
+      $scope.alerts = {
```
One of these should have a `details` property that we test

spadgett commented:
@jwforres any objections to moving things under `app/components` as we convert these?

jwforres commented:
thats fine, we just need to make sure grunt is doing the rights thing for both build and serve

TheRealJon commented:
@benjaminapetersen @spadgett @jwforres I believe I have addressed all of the comments. I moved the alerts component, template, and spec test to be collocated in `app/components` and updated grunt build and serve, as well as karma config. I also removed the unused properties in the alert component, and updated the rest to one-way data binding. Could someone take a close look at the grunt file to make sure I did everything that was needed there? I'm not intimately familiar with the build process yet and just want to make sure I didn't miss anything. Just looked back over the comments and realized I missed the one about the alert details property. I will add a test in there for that.

spadgett commented:
@TheRealJon Thanks, you should be able to run `grunt serve:dist` to test what's actually built.

TheRealJon commented:
I added a check for the details property of an alert and also added it to the comments in the alerts component. Also, ran `grunt server:dist` just to make sure the build was working correctly and everything seems to work.

benjaminapetersen commented on app/components/alerts/alerts.component.spec.js:
```diff
@@ -0,0 +1,126 @@
+"use strict";
+
```
Also, I think I do like having the test co-located with the file.  If we improve the structure so that related bits are all together, it makes it very clear when something isn't properly tested.

TheRealJon commented:
@spadgett @jwforres What do you think? Can I remove WIP status? Just trying to nudge this through before the end of the week since I will be gone for 3 weeks. 😄

spadgett commented:
@TheRealJon were you able to try `grunt serve:dist`?

TheRealJon commented:
@spadgett Yes, I added a test alert to the overview controller and ran `grunt serve:dist` and it seemed to work fine.

TheRealJon commented on test/karma.conf.js:
```diff
@@ -102,11 +102,13 @@ module.exports = function(config) {
       "bower_components/origin-web-catalog/dist/vendor-bundle.js",
       "bower_components/origin-web-catalog/dist/origin-web-catalogs.js",
       'app/scripts/**/*.js',
-      'app/views/directives/**/*.html',
```
We are not using any of the templates in that directory for our tests.

TheRealJon replied:
As long as we plan on moving things to a new component based file structure and writing tests for our components/directives as we do so, we shouldn't need that glob pattern any more.

jwforres commented:
Ok, I don't have a strong opinion here. Interested in Ben and Sam's
thoughts.

On Dec 11, 2017 4:12 PM, "Jon Jackson" <notifications@github.com> wrote:

> As far as the file naming convention, I chose this pattern for a few
> reasons:
>
>    1. It's a convention I have seen used in other places.
>    2. It makes the purpose/content of each file obvious (for both js and
>    html files). If we had an alerts service and an alerts controller, then we
>    would have a total of 6 js files in that folder including spec tests, so
>    they need to be differentiated. Naming the template in the same way is more
>    to hold the convention than anything else.
>    3. It makes writing glob patterns easier to get all
>    components/controllers/templates/etc. (For example
>    app/**/*.template.html will not give you index.html, or any other html file
>    that may exist but is not a template).
>    4. (# 3) would be handy If we break out another folder like "common"
>    where components that don't relate to any specific area of the app would
>    live. Alerts would actually be a good example of something that would
>    belong there, but since it's the first component, I think it's okay where
>    it is.
>    5. Similar reason for using fuzzy search in text editors/ide's
>
> I don't feel that strongly about it, but I think it's worth doing.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/2567#issuecomment-350859310>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7RiQXF3ADnmoHiKu2LgFTiXBFKvgks5s_ZptgaJpZM4Q227v>
> .
>

benjaminapetersen commented:
I love sharing opinions about things that don't really matter 😜  

So my initial thought is that since everything is under `/components` as a directory, having the extension be `.component.js` is probably redundant.  Template files with `.template.html` seem a bit lengthy as well, unless we have `.html` files that are not templates. (partials?) 

But I do tend to like [this style guide](https://github.com/toddmotto/angularjs-styleguide), which gives the following example for an angular 1.6 app:

```
/components/calendar
- calendar.module.js
- calendar.component.js
- calendar.service.js
- calendar.directive.js
- calendar.filter.js
- calendar.spec.js
- calendar.html
- calendar.scss
```

So perhaps more structure is better than less?  I'm sure nobody wants to rename all these files later.

(I think I very helpfully did not answer the question)

TheRealJon commented:
I can concede the `template.html` part because it really isn't all that useful (I just did it for consistency), but I still think the `.component.js` part would be helpful. The `components` folder in this case represents a broader context (something like "application components") while the file name implies more specifically that this file is an Angular Component. Plus if we are going to have all of the js files for an application component collocated, they will need to be differentiated somehow. Does anyone have any alternative ideas for this?

benjaminapetersen commented:
Yup, I think that's exactly where I land.

jwforres commented:
Agreed, the .template.html is really the only one that bothered me.

On Tue, Dec 12, 2017 at 2:05 PM, Ben Petersen <notifications@github.com>
wrote:

> Yup, I think that's exactly where I land.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-console/pull/2567#issuecomment-351156979>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7WOa0-KdI4J9lFbg2LMRE2WyccUxks5s_s6FgaJpZM4Q227v>
> .
>

TheRealJon commented:
@jwforres @spadgett @benjaminapetersen File names fixed, should I squash then remove WIP?

jwforres commented:
@TheRealJon yes please do

TheRealJon commented:
All done! Someone say the magic word! 😄 
@jwforres @spadgett

jwforres commented:
/lgtm
